File: go/dev_console.go
package dexcost

import (
	"fmt"
	"os"

	"github.com/shopspring/decimal"

	"github.com/DexwoxBusiness/dexcost-go/core"
)

// devMode controls whether development mode console output is enabled.
// When true, every recorded event is printed to stderr with a formatted summary.
var devMode bool

// IsDevMode returns true when development mode is active.
func IsDevMode() bool {
	return devMode
}

// EnableDevMode enables development mode output. All events and task completions
// are printed to stderr. Cloud sync is disabled.
func EnableDevMode() {
	devMode = true
	devPrint("dev mode \u2014 cloud sync disabled")
}

// disableDevMode resets development mode. Exported only for testing.
func disableDevMode() {
	devMode = false
}

// LogEvent prints a single event to stderr when dev mode is active.
func LogEvent(event *core.Event, taskType string) {
	if !devMode {
		return
	}

	cost := event.CostUSD
	confidence := event.CostConfidence

	switch event.EventType {
	case core.EventTypeLLMCall:
		provider := event.Provider
		if provider == "" {
			provider = "?"
		}
		model := event.Model
		if model == "" {
			model = "?"
		}
		inTok := 0
		if event.InputTokens != nil {
			inTok = *event.InputTokens
		}
		outTok := 0
		if event.OutputTokens != nil {
			outTok = *event.OutputTokens
		}
		cached := 0
		if event.CachedTokens != nil {
			cached = *event.CachedTokens
		}
		retryTag := ""
		if event.IsRetry {
			retryTag = "  \033[33m(retry)\033[0m"
		}
		cacheTag := ""
		if cached > 0 {
			cacheTag = fmt.Sprintf("  cached: %d", cached)
		}
		devPrint(fmt.Sprintf(
			"\033[32m+\033[0m llm_call  %s/%s  %d in / %d out%s  $%s%s%s",
			provider, model, inTok, outTok, cacheTag, cost.String(), retryTag, taskTag(taskType),
		))

	case core.EventTypeExternalCost, core.EventTypeComputeCost:
		service := event.ServiceName
		if service == "" {
			service = "unknown"
		}
		if confidence == core.CostConfidenceUnknown || cost.Equal(decimal.Zero) {
			devPrint(fmt.Sprintf(
				"\033[33m!\033[0m %s  %s  $0.00 \033[33m(no rate configured)\033[0m%s",
				string(event.EventType), service, taskTag(taskType),
			))
		} else {
			devPrint(fmt.Sprintf(
				"\033[32m+\033[0m %s  %s  $%s%s",
				string(event.EventType), service, cost.String(), taskTag(taskType),
			))
		}

	case core.EventTypeRetryMarker:
		reason := event.RetryReason
		if reason == "" {
			reason = "unknown"
		}
		devPrint(fmt.Sprintf(
			"\033[33m~\033[0m retry_marker  reason: %s  $%s%s",
			reason, cost.String(), taskTag(taskType),
		))
	}
}

// LogTaskComplete prints a task completion summary to stderr when dev mode is active.
func LogTaskComplete(task *core.Task) {
	if !devMode {
		return
	}

	retryInfo := ""
	if task.RetryCount > 0 {
		retryInfo = fmt.Sprintf("  retries: %d  retry cost: $%s",
			task.RetryCount, task.RetryCostUSD.String())
	}

	devPrint(fmt.Sprintf(
		"\033[36m+\033[0m task %s  %s  total: $%s%s",
		string(task.Status), task.TaskType, task.TotalCostUSD.String(), retryInfo,
	))
}

// taskTag returns a formatted task type tag for console output.
func taskTag(taskType string) string {
	if taskType != "" {
		return fmt.Sprintf("  \033[90m(task: %s)\033[0m", taskType)
	}
	return ""
}

// devPrint writes a formatted message to stderr with the dexcost prefix.
func devPrint(msg string) {
	fmt.Fprintf(os.Stderr, "\033[36m[dexcost]\033[0m %s\n", msg)
}

File: python/src/dexcost/dev_console.py
"""Development mode console output for dexcost.

When DEXCOST_ENV=development or environment="development" is passed to init(),
every recorded event is printed to stderr with a formatted summary.
"""

from __future__ import annotations

import sys
from decimal import Decimal
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dexcost.models.event import Event
    from dexcost.models.task import Task

# Module-level flag
_dev_mode: bool = False


def is_dev_mode() -> bool:
    return _dev_mode


def enable_dev_mode() -> None:
    global _dev_mode
    _dev_mode = True
    _print("\033[36m[dexcost]\033[0m dev mode — cloud sync disabled")


def log_event(event: "Event", task_type: str = "") -> None:
    """Print a single event to stderr."""
    if not _dev_mode:
        return

    cost = event.cost_usd
    confidence = event.cost_confidence

    if event.event_type == "llm_call":
        provider = event.provider or "?"
        model = event.model or "?"
        in_tok = event.input_tokens or 0
        out_tok = event.output_tokens or 0
        cached = event.cached_tokens or 0
        retry_tag = "  \033[33m(retry)\033[0m" if event.is_retry else ""
        cache_tag = f"  cached: {cached:,}" if cached > 0 else ""
        _print(
            f"\033[32m✓\033[0m llm_call  {provider}/{model}  "
            f"{in_tok:,} in / {out_tok:,} out{cache_tag}  "
            f"${cost}{retry_tag}"
            f"{_task_tag(task_type)}"
        )
    elif event.event_type in ("external_cost", "compute_cost"):
        service = event.service_name or "unknown"
        if confidence == "unknown" or cost == Decimal("0"):
            _print(
                f"\033[33m⚠\033[0m {event.event_type}  {service}  "
                f"$0.00 \033[33m(no rate configured)\033[0m"
                f"{_task_tag(task_type)}"
            )
        else:
            _print(
                f"\033[32m✓\033[0m {event.event_type}  {service}  "
                f"${cost}{_task_tag(task_type)}"
            )
    elif event.event_type == "retry_marker":
        reason = event.retry_reason or "unknown"
        _print(
            f"\033[33m↻\033[0m retry_marker  reason: {reason}  "
            f"${cost}{_task_tag(task_type)}"
        )


def log_task_complete(task: "Task") -> None:
    """Print task completion summary to stderr."""
    if not _dev_mode:
        return

    retry_info = ""
    if task.retry_count > 0:
        retry_info = (
            f"  retries: {task.retry_count}  "
            f"retry cost: ${task.retry_cost_usd}"
        )

    _print(
        f"\033[36m✓\033[0m task {task.status}  {task.task_type}  "
        f"total: ${task.total_cost_usd}{retry_info}"
    )


def _task_tag(task_type: str) -> str:
    if task_type:
        return f"  \033[90m(task: {task_type})\033[0m"
    return ""


def _print(msg: str) -> None:
    print(f"\033[36m[dexcost]\033[0m {msg}", file=sys.stderr)
